[WIP] Check block headers

zer0main opened the pull request.

zer0main committed:
Rename verify.go to client.go

More general logic will be added to the file,
so we need more general name.

zer0main committed:
sialiteclient: header -> headerBytes

Name 'header' will be needed later on for the BlockHeder variable.

zer0main committed:
Implement dummy cache.VerifyBlockHeader, call it from sialiteclient

zer0main committed:
Add .gitignore

zer0main committed:
VerifyBlockHeader: add extreme future check for timestamps

zer0main committed:
VerifyBlockHeader: add near future check

zer0main committed:
Gofmt

zer0main committed:
Implement getHeadersSlice()

zer0main committed:
Implement minimumValidChildTimestamp()

zer0main committed:
verifyBlockHeader: add past check, don't call it from sialiteclient yet

zer0main committed:
Implement VerifyBlockHeaders()

zer0main committed:
Call VerifyBlockHeaders() from sialiteclient

starius commented on cache/client.go:
```diff
@@ -0,0 +1,122 @@
+
+func verifyBlockHeader(header types.BlockHeader, minTimestamp types.Timestamp) error {
+	// Check that the timestamp is not too far in the past to be acceptable.
+	if minTimestamp > header.Timestamp {
```
Style: `header.Timestamp < minTimestamp`.
(Putting checked value as first operand increases readability.)

zer0main replied:
Done.

starius commented on cache/client.go:
```diff
@@ -0,0 +1,122 @@
+	// future and extreme future because there is an assumption that by the time
+	// the extreme future arrives, this block will no longer be a part of the
+	// longest fork because it will have been ignored by all of the miners.
+	if header.Timestamp > types.CurrentTimestamp()+types.ExtremeFutureThreshold {
```
Can you pass CurrentTimestamp as an argument? It would make testing easier.

zer0main replied:
No, because the time passes between checks (`ExtremeFutureThreshold` and `FutureThreshold` checks). The main check (of target) will be added later on between those two checks, and the computation takes some time.

starius replied:
In our case we already downloaded all headers from this node so the if it want to feed us with garbage it already succeeds. We need to check >150000 blocks very fast, so the time we spend on one block header should be really small (in Sia not only the header is checked so the time matters). I suggest to remove extra check and extra (possibly expensive) call to `CurrentTimestamp()`

starius replied:
Just add a comment that check against types.ExtremeFutureThreshold is omitted because it does included in the further check and does not make sense as optimization in our case.

zer0main replied:
Done.

starius commented on cache/client.go:
```diff
@@ -0,0 +1,122 @@
+
+	// Check if the block is in the near future, but too far to be acceptable.
+	// This is the last check because it's an expensive check, and not worth
+	// performing if the earlier checks failed.
```
I don't see how this check is more expensive than the previous one. Maybe we can remove the previous check if it always fails when this check fails?

zer0main replied:
See the updates comment and my response above.

starius commented on cache/client.go:
```diff
@@ -0,0 +1,122 @@
+	for i := 1; i < int(types.MedianTimestampWindow); i++ {
+		// If the genesis block is 'parent', use the genesis block timestamp
+		// for all remaining times.
+		if parent == (types.BlockID{}) {
```
s/(types.BlockID{})/types.BlockID{}/

zer0main replied:
Won't work out because of `if`, we get syntax error without parentheses.

starius replied:
I see. `{` is attracted to `if` more than to types.BlockID

starius commented on cache/client.go:
```diff
@@ -0,0 +1,122 @@
+			continue
+		}
+
+		if (headerIndex - i) < 0 {
```
`if headerIndex - i < 0 {` or `if headerIndex < i {`

zer0main replied:
Done.

starius commented on cache/client.go:
```diff
@@ -0,0 +1,122 @@
+	parent := headers[headerIndex].ParentID
+	for i := 1; i < int(types.MedianTimestampWindow); i++ {
+		// If the genesis block is 'parent', use the genesis block timestamp
+		// for all remaining times.
```
Maybe we can move this case outside of the loop and just return genesis block timestamp early?

zer0main replied:
This isn't just for genesis case itself, but for all the blocks which have genesis block less than types.MedianTimestampWindow times behind. This is the computation rule for them.

starius replied:
```
if number of blocks < types.MedianTimestampWindow {
  return timestamp of genesis block
}
loop
```

zer0main replied:
The algorithm here is different, we add the actual timestamps of blocks till genesis block, after that we fill the remaining spots (if any) with GenesisTimestamp till `MedianTimestampWindow` length. We finally return the median of this list, which is not necessary equal to the timestamp of genesis block.

starius replied:
I see.

starius commented on cache/client.go:
```diff
@@ -0,0 +1,122 @@
+}
+
+func getHeadersSlice(headers []byte) (headersSlice []types.BlockHeader, err error) {
+	headersSlice = make([]types.BlockHeader, len(headers))
```
Probably `len(headers)/48`

zer0main replied:
Fixed.

starius commented on cache/client.go:
```diff
@@ -0,0 +1,122 @@
+	return windowTimes[len(windowTimes)/2], nil
+}
+
+func getHeadersSlice(headers []byte) (headersSlice []types.BlockHeader, err error) {
```
I suggest to remove names of result variables, since they are obvious from function name and their types.

starius commented on cache/client.go:
```diff
@@ -0,0 +1,122 @@
+
+// minimumValidChildTimestamp returns the earliest timestamp that a child node
+// can have while still being valid.
+func minimumValidChildTimestamp(headers []types.BlockHeader, headerIndex int) (minTimestamp types.Timestamp, err error) {
```
I suggest to remove names of result variables, since they are obvious from function name and their types and not used in function body.

starius commented on cache/client.go:
```diff
@@ -0,0 +1,122 @@
+		Timestamp:  types.GenesisTimestamp,
+		MerkleRoot: types.GenesisBlock.MerkleRoot(),
+	}
+	for i := 1; i < len(headers); i++ {
```
Something looks wrong here. You iterate by bytes, but one header takes 48 bytes.

zer0main replied:
Fixed.

starius commented on cache/client.go:
```diff
@@ -0,0 +1,122 @@
+		MerkleRoot: types.GenesisBlock.MerkleRoot(),
+	}
+	for i := 1; i < len(headers); i++ {
+		headersSlice[i] = types.BlockHeader{
```
Can you first create a sub-slice of just this header (a slice of 48 bytes)? I think it will simplify the further extractions.

zer0main replied:
Done.

zer0main replied:
By the way, is the layout here the same like in Sia's `BlockHeader` (because this is how I consider it here)?
`0-8` bytes: Nonce
`8-16` bytes: Timestamp
`16-48` bytes: MerkleRoot

starius replied:
[Yes](https://github.com/starius/sialite/blob/64ebbbf3f0e534452ad8adaa6c3febbd40c0e773/cache/build.go#L38) and it is intended.

starius commented on cache/client.go:
```diff
@@ -0,0 +1,122 @@
+		copy(headersSlice[i].Nonce[:], headers[i*48:(i*48+8)])
+		copy(headersSlice[i].MerkleRoot[:], headers[(i*48+16):(i*48+48)])
+	}
+	if len(headers) > 1 {
```
len(headers)/48 ?

starius commented on cache/client.go:
```diff
@@ -0,0 +1,122 @@
+		copy(headersSlice[i].MerkleRoot[:], headers[(i*48+16):(i*48+48)])
+	}
+	if len(headers) > 1 {
+		if headersSlice[1].ParentID != types.GenesisID {
```
Combine such `if`s:
`if len(headers) > 1 && headersSlice[1].ParentID != types.GenesisID {`

zer0main replied:
Done.

starius commented on cache/client.go:
```diff
@@ -0,0 +1,122 @@
+	if err != nil {
+		return err
+	}
+	if len(headers) < 1 {
```
== 0

zer0main replied:
Done.

starius commented on cache/client.go:
```diff
@@ -0,0 +1,122 @@
+	if len(headers) < 1 {
+		return fmt.Errorf("Can't verify list of 0 headers")
+	}
+	minTimestamp := headersSlice[0].Timestamp
```
Doesn't minimumValidChildTimestamp return headersSlice[0].Timestamp for the first block? We could avoid this special case, move this variable inside the loop and run minimumValidChildTimestamp before verifyBlockHeader.

zer0main replied:
The point is that we need to call it with parent of the current node (`i-1` in our case). But we must iterate starting with 0 in order to check 0th header, so we would need to call `minimumValidChildTimestamp` with -1 as index, which is not correct.

zer0main replied:
Don't think it's worth it. We don't lose anything by doing this way.

starius reviewed (changes requested):
Thanks! Just few comments.

zer0main committed:
Fixes

zer0main committed:
Remove result var names from function signatures

zer0main committed:
Remove ExtremeFuture check

zer0main committed:
Remove headerIndex arg of minimumValidChildTimestamp

zer0main commented on cache/client.go:
```diff
@@ -0,0 +1,122 @@
+
+// minimumValidChildTimestamp returns the earliest timestamp that a child node
+// can have while still being valid.
+func minimumValidChildTimestamp(headers []types.BlockHeader, headerIndex int) (minTimestamp types.Timestamp, err error) {
```
Yeah, it would. For more clear function signature.
PS
Implemented by now.

zer0main committed:
Add checkTarget() func

zer0main committed:
Implement calculateBlockTotals()

zer0main committed:
Implement oak difficulty adjustment algorithm

zer0main committed:
Add targetAdjustmentBase()

Will be used in old, broken difficulty adjustment algorithm.
(For < OakHardforkBlock blocks).

zer0main committed:
Implement clampTargetAdjustment()

Will be used in old, broken difficulty adjustment algorithm.
(For < OakHardforkBlock blocks).

zer0main committed:
Add oldTargetAdjustment()

This is old, broken difficulty adjustment algorithm.
Applies for blocks before OakHardforkBlock.

zer0main committed:
Implement calculateChildTarget()

This function uses either oak adjustment or old broken
adjustment algorithm, depending on block height.

zer0main committed:
Add getTargets()

zer0main committed:
verifyBlockHeader(): check the target

zer0main committed:
Fix getTargets(): last height must be len-1

zer0main committed:
Add few useful types

starius commented:
I rebased and merged this in https://github.com/starius/sialite/pull/2
Note: currently headers check fails on real blockchain with "Block header validation failed: Block is unsolved", so I changed it temporarily from error to warning.
